src/librustdoc/passes/strip_hidden.rs

```diff
@@ -36,14 +44,37 @@ pub(crate) fn strip_hidden(krate: clean::Crate, cx: &mut DocContext<'_>) -> clea
     stripper.fold_crate(krate)
 }
 
-struct Stripper<'a> {
+struct Stripper<'a, 'tcx> {
     retained: &'a mut ItemIdSet,
     update_retained: bool,
+    tcx: TyCtxt<'tcx>,
+    is_in_hidden_item: bool,
 }
 
-impl<'a> DocFolder for Stripper<'a> {
+impl<'a, 'tcx> DocFolder for Stripper<'a, 'tcx> {
     fn fold_item(&mut self, i: Item) -> Option<Item> {
-        if i.attrs.lists(sym::doc).has_word(sym::hidden) {
+        let is_hidden =
+            if matches!(*i.kind, clean::ImplItem(box clean::Impl { trait_: Some(_), .. })) {
+                // Unless directly hidden, trait implementations are considering public.
+                i.attrs.lists(sym::doc).has_word(sym::hidden)
+            } else {
+                let mut is_hidden =
+                    self.is_in_hidden_item || i.attrs.lists(sym::doc).has_word(sym::hidden);
+                if !is_hidden && i.inline_stmt_id.is_none() {
+                    let hir = self.tcx.hir();
+                    is_hidden = i
+                        .inline_stmt_id
+                        .or_else(|| i.item_id.as_def_id())
+                        .and_then(|def_id| def_id.as_local())
+                        .map(|def_id| {
+                            let hir_id = hir.local_def_id_to_hir_id(def_id);
+                            inherits_doc_hidden(self.tcx, hir_id)
+                        })
+                        .unwrap_or(false);
+                }
+                is_hidden
+            };
```

Comment:
This code now defines how `rustdoc` handles `doc(hidden)` item and its children. So to sum it up:

 If it's an impl item, then unless there is `doc(hidden)` directly on it, we still keep it, otherwise we discard everything.

There is a little twist: if it's an item that was reexported, then we use the reexport `DefId` to check if it's inside a `doc(hidden)` item.

Reply:
> If it's an impl item, then unless there is doc(hidden) directly on it, we still keep it

Even if either the implemented trait or the implementing type is hidden, too? That feels incorrect. Consider:

```rust
#[doc(hidden)]
const _: () = {
    pub trait Tr {}
    impl Tr for Ty {} // this impl should be hidden, right?
};

pub struct Ty;
```

```rust
#[doc(hidden)]
const _: () = {
    pub struct Ty;
    impl Tr for Ty {} // this impl should be hidden, right?
};

pub trait Tr {}
```

Reply:
So I remove this special case and make everyone happy? :)

Reply:
Simplified it then, no more exceptions.